File: Data Structures & Algorithms/validate-parentheses/submission-7.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for char in s:
            if char in ('(', '{', '['): stack.append(char)
            else:
                if not stack: return False
                elif char == ')':
                    if stack.pop() != '(': return False
                elif char == '}':
                    if stack.pop() != '{': return False
                elif char == ']':
                    if stack.pop() != '[': return False
        return not stack

File: Data Structures & Algorithms/minimum-stack/submission-7.java
class MinStack {

    private int min;
    private Deque<List<Integer>> stack;

    public MinStack() {
        stack = new ArrayDeque<>();
    }
    
    public void push(int val) {
        int currMin = stack.size() == 0 ? val : Math.min(stack.peekFirst().get(1), val);
        stack.push(List.of(val, currMin));
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peekFirst().get(0);
    }
    
    public int getMin() {
        return stack.peekFirst().get(1);
    }
}
